govuk_chat_evaluation/question_router/evaluate.py

```diff
@@ -0,0 +1,183 @@
+    accuracy_score,
+    precision_score,
+    recall_score,
+    f1_score,
+    fbeta_score,
+    confusion_matrix,
+)
+from tabulate import tabulate
+import matplotlib.pyplot as plt
+import seaborn as sns
+
+from ..file_system import jsonl_to_models, write_csv_results
+
+
+class EvaluationResult(BaseModel):
+    question: str
+    expected_outcome: str
+    actual_outcome: str
+    confidence_score: float
+
+    def for_csv(self) -> dict[str, Any]:
+        return {**self.model_dump()}
+
+
+class AggregateResults:
+    def __init__(self, evaluation_results: list[EvaluationResult]):
+        self.evaluation_results = evaluation_results
+
+    @cached_property
+    def classification_labels(self) -> list[str]:
+        return sorted(
+            list(set(item for lst in self._expected_actual_lists for item in lst))
+        )
+
+    @cached_property
+    def _expected_actual_lists(self) -> tuple[list[str], list[str]]:
+        pairs_list = [
+            (eval.expected_outcome, eval.actual_outcome)
+            for eval in self.evaluation_results
+        ]
+
+        expected, actual = zip(*pairs_list)
+        return list(expected), list(actual)
+
+    def accuracy(self) -> float:
+        return accuracy_score(
+            *self._expected_actual_lists,  # type: ignore
+        )
+
+    def precision(self) -> float:
+        return precision_score(
+            *self._expected_actual_lists,
+            average="weighted",
+            zero_division=0,  # type: ignore
+        )
+
+    def recall(self) -> float:
+        return recall_score(
+            *self._expected_actual_lists,
+            average="weighted",
+            zero_division=0,  # type: ignore
+        )
+
+    def f1_score(self) -> float:
+        return f1_score(
+            *self._expected_actual_lists,
+            average="weighted",
+            zero_division=0,  # type: ignore
+        )
+
+    def f2_score(self) -> float:
+        return fbeta_score(
+            *self._expected_actual_lists,
+            beta=2,
+            average="weighted",
+            zero_division=0,  # type: ignore
+        )
+
+    def confusion_matrix_data(self) -> list[list[int]]:
+        return confusion_matrix(
+            *self._expected_actual_lists,
+            labels=sorted(list(set(self.classification_labels))),  # type: ignore
+        )
+
+    def miscategorised_cases(self) -> list[dict[str, Any]]:
+        miscategorised = [
+            result
+            for result in self.evaluation_results
+            if result.expected_outcome != result.actual_outcome
+        ]
+
+        return [
+            {
+                "question": miscategorised[i].question,
+                "predicted_classification": miscategorised[i].expected_outcome,
+                "actual_classification": miscategorised[i].actual_outcome,
+                "confidence_score": miscategorised[i].confidence_score,
+            }
+            for i in range(len(miscategorised))
```

Comment:
could this be `for item in miscategorised` ? 

I also wonder whether you need two loops and could do:

```
        return [
            {
                "question": result.question,
                "predicted_classification": result.expected_outcome,
                "actual_classification": result.actual_outcome,
                "confidence_score": result.confidence_score,
            }
            for result in self.evaluation_results
            if result.expected_outcome != result.actual_outcome
    ]
```